engine/src/main/java/org/pentaho/di/core/util/MetaGenerator.java

```diff
@@ -0,0 +1,225 @@
+    new TypeFieldDefinition( ValueMetaInterface.TYPE_STRING, "LocalityField" ),
+    new TypeFieldDefinition( ValueMetaInterface.TYPE_STRING, "PostTownField" ),
+    new TypeFieldDefinition( ValueMetaInterface.TYPE_STRING, "DeliveryServiceQualifierField" ),
+    new TypeFieldDefinition( ValueMetaInterface.TYPE_STRING, "PostalCodeField" ),
+    new TypeFieldDefinition( ValueMetaInterface.TYPE_STRING, "SubAdministrativeAreaField" ),
+    new TypeFieldDefinition( ValueMetaInterface.TYPE_STRING, "AdministrativeAreaField" ),
+    new TypeFieldDefinition( ValueMetaInterface.TYPE_STRING, "CountryNameField" ),
+    new TypeFieldDefinition( ValueMetaInterface.TYPE_STRING, "CountryCodeField" ), };
+
+  private TypeFieldDefinition[] fields;
+
+  public MetaGenerator( TypeFieldDefinition[] fields ) {
+    this.fields = fields;
+  }
+
+  public String generateCode() {
+
+    StringBuilder code = new StringBuilder( 5000 );
+
+    // Generate the declarations...
+    //
+    for ( TypeFieldDefinition field : fields ) {
+      code.append( "  private " + field.getTypeDescription() + " " + field.getMemberName() + ";" ).append(
+        Const.CR );
+    }
+    code.append( Const.CR );
+
+    // Generate getXML()...
+    //
+    code.append( "  public String getXML() throws KettleException {" ).append( Const.CR );
+    code.append( "    StringBuilder xml = new StringBuilder(100);" ).append( Const.CR );
+    for ( TypeFieldDefinition field : fields ) {
+      code.append(
+        "    xml.append(XMLHandler.addTagValue(\""
+          + field.getFieldName() + "\", " + field.getMemberName() + "));" ).append( Const.CR );
+    }
+    code.append( "    return xml.toString();" ).append( Const.CR );
+    code.append( "  }" ).append( Const.CR );
+    code.append( Const.CR );
+
+    // Generate loadXML()...
+    //
+    code.append( "  public void loadXML(Node stepnode, List<DatabaseMeta> databases, "
+      + "IMetaStore metaStore) throws KettleXMLException {" ).append( Const.CR );
+    for ( TypeFieldDefinition field : fields ) {
+      switch ( field.getType() ) {
+        case ValueMetaInterface.TYPE_STRING:
+          code.append(
+            "    "
+              + field.getMemberName() + " = XMLHandler.getTagValue(stepnode, \"" + field.getFieldName()
+              + "\");" ).append( Const.CR );
+          break;
+        case ValueMetaInterface.TYPE_BOOLEAN:
+          code.append(
+            "    "
+              + field.getMemberName() + " = \"Y\".equalsIgnoreCase(XMLHandler.getTagValue(stepnode, \""
+              + field.getFieldName() + "\"));" ).append( Const.CR );
+          break;
+        case ValueMetaInterface.TYPE_INTEGER:
+          code.append(
+            "    "
+              + field.getMemberName() + " = Const.toInt(XMLHandler.getTagValue(stepnode, \""
+              + field.getFieldName() + "\"), -1);" ).append( Const.CR );
+          break;
+        default:
+          break;
+      }
+    }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 89:112, 134:157.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.